addrmgr/addrmanager.go

```diff
@@ -919,79 +879,93 @@ func (a *AddrManager) GetAddress() *KnownAddress {
-			a.addrNew[subnetworkID], a.nNew[subnetworkID])
-	}
+	addrTriedBucket := a.addrTriedBucket(a.localSubnetworkID)
+	nTriedNodes := a.nTriedNodes(a.localSubnetworkID)
+	addrNewBucket := a.addrNewBucket(a.localSubnetworkID)
+	nNewNodes := a.nNewNodes(a.localSubnetworkID)
+	knownAddress := a.getAddress(addrTriedBucket, nTriedNodes, addrNewBucket, nNewNodes)
 
 	return knownAddress
 
 }
 
-// see GetAddress for details
+// getAddress returns a single address that should be routable.
+// See GetAddress for further details.
 func (a *AddrManager) getAddress(addrTried *triedBucket, nTried int, addrNew *newBucket, nNew int) *KnownAddress {
 	// Use a 50% chance for choosing between tried and new table entries.
+	var addrBucket bucket
 	if nTried > 0 && (nNew == 0 || a.rand.Intn(2) == 0) {
-		// Tried entry.
-		large := 1 << 30
-		factor := 1.0
-		for {
-			// pick a random bucket.
-			bucket := a.rand.Intn(len(addrTried))
-			if addrTried[bucket].Len() == 0 {
-				continue
-			}
+		addrBucket = addrTried
+	} else if nNew > 0 {
+		addrBucket = addrNew
+	} else {
+		// There aren't any addresses in any of the buckets
+		return nil
+	}
 
-			// Pick a random entry in the list
-			e := addrTried[bucket].Front()
-			for i :=
-				a.rand.Int63n(int64(addrTried[bucket].Len())); i > 0; i-- {
-				e = e.Next()
-			}
-			ka := e.Value.(*KnownAddress)
-			randval := a.rand.Intn(large)
-			if float64(randval) < (factor * ka.chance() * float64(large)) {
-				log.Tracef("Selected %s from tried bucket",
-					NetAddressKey(ka.na))
-				return ka
-			}
-			factor *= 1.2
+	// Pick a random bucket
+	randomBucket := addrBucket.randomBucket(a.rand)
+
+	// Get the sum of all chances
+	totalChance := float64(0)
+	for _, ka := range randomBucket {
+		totalChance += ka.chance()
+	}
+
+	// Pick a random address weighted by chance
+	randomValue := a.rand.Float64()
+	accumulatedChance := float64(0)
+	for _, ka := range randomBucket {
+		normalizedChance := ka.chance() / totalChance
+		accumulatedChance += normalizedChance
+		if randomValue < accumulatedChance {
+			return ka
 		}
-	} else if nNew > 0 {
-		// new node.
-		// XXX use a closure/function to avoid repeating this.
-		large := 1 << 30
-		factor := 1.0
-		for {
-			// Pick a random bucket.
-			bucket := a.rand.Intn(len(addrNew))
-			if len(addrNew[bucket]) == 0 {
-				continue
-			}
-			// Then, a random entry in it.
-			var ka *KnownAddress
-			nth := a.rand.Intn(len(addrNew[bucket]))
-			for _, value := range addrNew[bucket] {
-				if nth == 0 {
-					ka = value
-				}
-				nth--
-			}
-			randval := a.rand.Intn(large)
-			if float64(randval) < (factor * ka.chance() * float64(large)) {
-				log.Tracef("Selected %s from new bucket",
-					NetAddressKey(ka.na))
-				return ka
-			}
-			factor *= 1.2
+	}
+
+	panic("randomValue is exactly 1, which cannot happen")
+}
+
+type bucket interface {
+	name() string
+	randomBucket(random *rand.Rand) []*KnownAddress
```

Comment:
It's quite weird that randomBucket doesn't return a bucket

Reply:
We spoke about this and decided to leave it as is.